### PR TITLE
Prevent a null ref exception when device does not specify a desired p…

### PR DESCRIPTION
**Pull request by drwill-ms**
Prevent a null ref exception when device does not specify a desired prop callback

### CODE REVIEW

File: e2e/test/TwinE2ETests.cs [hunk 1]
```diff
@@ -202,10 +202,35 @@ public async Task Twin_ClientHandlesRejectionInvalidPropertyName_AmqpWs()
             await Twin_ClientHandlesRejectionInvalidPropertyName(Client.TransportType.Amqp_WebSocket_Only).ConfigureAwait(false);
         }
 
+        [DataTestMethod]
+        [DataRow(Client.TransportType.Amqp_Tcp_Only)]
+        [DataRow(Client.TransportType.Amqp_WebSocket_Only)]
+        public async Task Twin_ClientSetsReportedPropertyWithoutDesiredPropertyCallback(Client.TransportType transportType)
+        {
+            // arrange
+
```
Comment (abhipsaMisra) [inline e2e/test/TwinE2ETests.cs]: nit: extra line

File: iothub/device/src/InternalClient.cs [hunk 2]
```diff
@@ -1238,6 +1242,11 @@ public Task UpdateReportedPropertiesAsync(TwinCollection reportedProperties, Can
         //  Codes_SRS_DEVICECLIENT_18_005: When a patch is received from the service, the `callback` shall be called.
         internal void OnReportedStatePatchReceived(TwinCollection patch)
         {
+            if (this.desiredPropertyUpdateCallback == null)
```
Comment (davilu) [inline iothub/device/src/InternalClient.cs]: `this.desiredPropertyUpdateCallback?.Invoke(patch, this.twinPatchCallbackContext);`

File: e2e/test/TwinE2ETests.cs [hunk 3]
```diff
@@ -202,10 +202,35 @@ public async Task Twin_ClientHandlesRejectionInvalidPropertyName_AmqpWs()
             await Twin_ClientHandlesRejectionInvalidPropertyName(Client.TransportType.Amqp_WebSocket_Only).ConfigureAwait(false);
         }
 
+        [DataTestMethod]
```
Comment (davilu) [inline e2e/test/TwinE2ETests.cs]: Invalid test. Need to check connection status changes
```
       [TestMethod]
        public async Task Twin_UpdateDesiredPropertyWithoutCallback_Amqp()
        {
            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix).ConfigureAwait(false);
            using (var deviceClient = DeviceClient.CreateFromConnectionString(testDevice.ConnectionString, Client.TransportType.Amqp_Tcp_Only))
            {
                await Twin_DeviceSetsReportedPropertyAndGetsItBack(deviceClient).ConfigureAwait(false);

                int connectionStatusChangeCount = 0;
                ConnectionStatusChangesHandler connectionStatusChangesHandler = (ConnectionStatus status, ConnectionStatusChangeReason reason) =>
                {
                    Interlocked.Increment(ref connectionStatusChangeCount);
                };

                deviceClient.SetConnectionStatusChangesHandler(connectionStatusChangesHandler);

                var propName = Guid.NewGuid().ToString();
                var propValue = Guid.NewGuid().ToString();
                await RegistryManagerUpdateDesiredPropertyAsync(testDevice.Id, propName, propValue).ConfigureAwait(false);
                await Task.Delay(TimeSpan.FromSeconds(10)).ConfigureAwait(false);
                Assert.AreEqual(0, connectionStatusChangeCount, "AMQP should be disconnected.");
                await deviceClient.CloseAsync().ConfigureAwait(false);
            }
        }
```

**Pull request merged**